boards: stm32f3_disco: Add I2C support

ydamigos opened the pull request:
Add I2C support to STM32F3DISCOVERY board.

This PR also adds support for selecting the source
of STM32F0/3 I2C's independent clock.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
boards: stm32f3_disco: Add I2C support

Add I2C support to STM32F3DISCOVERY board

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

galak reviewed (changes requested):
Can we separate this into 2 patches, one for adding F3 support to the I2C driver, the other that adds the board support

ydamigos commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -103,6 +103,21 @@ static int i2c_stm32_init(struct device *dev)
+#elif defined(CONFIG_I2C_3)
+	LL_RCC_SetI2CClockSource(LL_RCC_I2C3_CLKSOURCE_SYSCLK);
+#endif
+
```
This is wrong. I will fix it.

erwango replied:
Which part?

ydamigos commented:
Yes I will separate it.

ydamigos committed:
drivers: i2c_ll_stm32: STM32F0/3 set I2C clock source

STM32F0/3 I2C's independent clock source supports only
HSI and SYSCLK, not APB1. We force I2C clock source to SYSCLK.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

dbkinder commented on boards/arm/stm32f3_disco/doc/stm32f3_disco.rst:
```diff
@@ -133,6 +137,18 @@ Serial Port
+I2C
+===
+
+STM32F3DISCOVERY has up to 2 I2Cs. I2C1 is connected to the LSM303DLHC is
```
Add an "and" before the last "is" on this line

ydamigos replied:
done

dbkinder reviewed (dismissed):
One little change

ydamigos commented:
@ldts Could you also take a look?

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -103,6 +103,33 @@ static int i2c_stm32_init(struct device *dev)
+	switch ((u32_t)cfg->i2c) {
+#ifdef CONFIG_I2C_1
+	case CONFIG_I2C_1_BASE_ADDRESS:
+		LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_SYSCLK);
```
We should take care of Clock control configuration.
If board is running on HSI, this won't work.

ydamigos replied:
I was just looking into the reference manuals. All SOCs with I2C v2 have an I2C independent clock. I will make the commit generic.

erwango replied:
Ok for APB1.
But looking to F334 and F303 ref manuals, in RCC chapters, clock tree diagrams show that I2Cx could be feed with HSI or SYSCLK.

ydamigos replied:
My last push checks if IS_ENABLED(CONFIG_STM32_PLL_SRC_HSI) and sets the clock accordingly.
```
if (IS_ENABLED(CONFIG_STM32_PLL_SRC_HSI)) {
	LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_HSI);
} else {
	LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_SYSCLK);
}
```

erwango reviewed (changes requested):
Take clock control configuration into account

ydamigos commented:
I update the commit to handle the I2C independent clock source for both F3 and F0 families.

erwango commented on boards/arm/stm32f3_disco/doc/stm32f3_disco.rst:
```diff
@@ -119,6 +121,8 @@ Default Zephyr Peripheral Mapping:
 - LD9 : PE12
 - LD10 : PE13
+- LSM303DLHC_I2C_SCL : PB6
+- LSM303DLHC_I2C_SDA : PB7
```
Maybe we should wait for this sensor to be supported before documenting its pinmapping on the board?

ydamigos replied:
removed

erwango commented on boards/arm/stm32f3_disco/stm32f3_disco_defconfig:
```diff
@@ -21,6 +21,10 @@ CONFIG_UART_STM32_PORT_1=y
 
+#enable I2C
+CONFIG_I2C=y
+CONFIG_I2C_STM32_V2=y
```
Can't we get this in SoC config?
For instance in arch/arm/soc/st_stm32/stm32f3/Kconfig.series:
select I2C_STM32_V2 if I2C

ydamigos replied:
done

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -103,6 +103,46 @@ static int i2c_stm32_init(struct device *dev)
+	switch ((u32_t)cfg->i2c) {
+#ifdef CONFIG_I2C_1
+	case CONFIG_I2C_1_BASE_ADDRESS:
+		if (IS_ENABLED(CONFIG_STM32_PLL_SRC_HSI)) {
```
I think this should be a Kconfig option.
User could chose I2C clock source to be either HSI or SYSCLK (with default SYSCLK).
If HSI, then you need to check if actually activated (in RCC registers, and not using Clock driver CONFIG_...)
if yes, select HSI
else trigger error
If SYSCLK (it should always be on, so it should be ok.)
 then select SYSCLK

ydamigos replied:
I agree with you that the driver should check the RCC registers and not the clock driver CONFIG.
I didn't find any LL function that checks if HSI is active. Is it ok if we read the register directly?
```READ_BIT(RCC->CR, RCC_CR_HSION) == (RCC_CR_HSION)```
I am skeptical about the option of selecting I2C clock from the Kconfig/dts file. It will be more confusing than the driver configuring it according to the configured system clock source.
If we decide to implement it using user input, we should to it for all families with independent I2C clock.

erwango replied:
> I didn't find any LL function that checks if HSI is active. Is it ok if we read the register directly?
READ_BIT(RCC->CR, RCC_CR_HSION) == (RCC_CR_HSION)

LL_RCC_HSI_IsReady is not strictly equivalent, but should provide the same service and will help you in code genericity

> I am skeptical about the option of selecting I2C clock from the Kconfig/dts file. It will be more confusing than the driver configuring it according to the configured system clock source.

Then, you'd only need to set to SYSCLK by default. Feeding I2C with HSI depending if HSI is activated or not make this uncertain and unflexible:
-Having HSI activated prevents you to have I2C feed by SYSCLK if you want to.
-You won't have same behavior depending on your HSI configuration (you could have requirements on HSI independant from I2C clocking).
For instance "Wakeup from Stop mode on address match" is available only when I2C clocked by HSI. What if you need HSI for other purpose but you don't want this feature.

> If we decide to implement it using user input, we should to it for all families with independent I2C clock.

Yes, this is only blocking for F3 right now, but we can say it should be available for all families. This could be done step by step though.
Other question would be : do we need to propose a clock selection API, as others IPs have the same constraint/options?

ydamigos replied:
I will set the I2C clock to SYSCLK for now. If there is a request/need I will come back to it.

erwango reviewed (changes requested):
Should be reworked with following:
-Take input clock as Kconfig option (or should it be dts property?)
-In case of HSI selected, check against RCC registers, and not clock control options (if not activated: trigger and error / activate it ? I'm not sure)
-Update i2c_stm32_runtime_configure to take into account HSI rate (8MHz) if selected

erwango commented:
Last point (to be confirmed):
In i2c_stm32_runtime_configure, in case selected clock is SYSCLK, clock_control_get_rate should take into account SYSCLK and not return APB BUS clock rate.

ydamigos commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -23,12 +23,34 @@ static int i2c_stm32_runtime_configure(struct device *dev, u32_t config)
+	switch ((u32_t)cfg->i2c) {
+#ifdef CONFIG_I2C_1
+	case CONFIG_I2C_1_BASE_ADDRESS:
+		clock = LL_RCC_GetI2CClockSource(LL_RCC_I2C1_CLKSOURCE);
```
This is wrong. I entered wrong function. I want to use LL_RCC_GetI2CClockFreq but the USE_FULL_LL_DRIVER is not defined.
@erwango which is the best way to define USE_FULL_LL_DRIVER?

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -23,12 +23,21 @@ static int i2c_stm32_runtime_configure(struct device *dev, u32_t config)
+	};
 
 	clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			(clock_control_subsys_t *) &pclken, &clock);
```
I'm affraid this won't work.
With this you'll get SystemCoreClock = core clock = AHB clock.
and here we'd like to get SYSCLK (SystemCoreClock * AHB prescaler).

ydamigos replied:
I thought that  SYSCLK was an abbreviation of SystemCoreClock. I should better study STM32 RCC. I will fix it.

ydamigos committed:
ext: stm32cube: Define USE_FULL_LL_DRIVER

This patch enables the LL API get funtions for system and
peripherals clocks frequency.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -23,12 +23,34 @@ static int i2c_stm32_runtime_configure(struct device *dev, u32_t config)
 
-	data->dev_config = config;
+#if defined(CONFIG_SOC_SERIES_STM32F3X) || defined(CONFIG_SOC_SERIES_STM32F0X)
+	switch ((u32_t)cfg->i2c) {
```
I feel this code would be nice if clock source source was a priori unknown, though, since it is not the case here, and hence it sounds a bit "over featured".
Isn't there an alternate solution to get SYSCLK value (whatever the instance)?

ydamigos replied:
We could use the LL_RCC_GetSystemClocksFreq(). I tried it but the code crashes inside this function.

ydamigos replied:
done

erwango reviewed (changes requested):
It would be nice to re-inforce message (code comments or commit message) that 
clock source should be HSI or SYSCLK and your implementation forces SYSCLK.

ydamigos commented:
@erwango I also updated code comments and commit message. Is it ok now?

erwango reviewed (commented):
Looks nice!
Did you had a chance to test it?

ydamigos commented:
It works on STM32F3Discovery revB board.

galak commented on ext/hal/st/stm32cube/Kbuild:
```diff
@@ -15,6 +15,8 @@ obj-y += stm32f3xx/drivers/src/stm32f3xx_hal.o
 obj-$(CONFIG_CLOCK_CONTROL_STM32_CUBE) += stm32f3xx/drivers/src/stm32f3xx_ll_utils.o
 obj-$(CONFIG_SERIAL_HAS_DRIVER) += stm32f3xx/drivers/src/stm32f3xx_hal_uart.o
+obj-$(CONFIG_I2C) += stm32f3xx/drivers/src/stm32f3xx_ll_i2c.o
+obj-$(CONFIG_I2C) += stm32f3xx/drivers/src/stm32f3xx_ll_rcc.o
```
I know we did this for the other STM32 I2C devices, but this really should be something like obj-$(CONFIG_I2C_STM32_V2) += ...

ydamigos replied:
CONFIG_I2C selects CONFIG_I2C_STM32_V2 in Kconfig.series. I don't really see any reason why we need to specify here the I2C controller version. If it is a must, I'll change it and I'll create a PR for the other families.
